pkg/restart/restart.go

```diff
@@ -0,0 +1,166 @@
+
+				debug("Start: context cancelled, finish signal received")
+
+				// the manager is not running anymore, now we can be sure of it. Let's record that fact in our state to keep the records
+				// straight.
+				debug("Start: context cancelled, about to set running=false")
+				s.lock.Lock()
+				debug("Start: locked")
+				s.running = false
+				s.lock.Unlock()
+				debug("Start: unlocked")
+				debug("Start: context cancelled, running=false is set")
+
+				// now, let's see what caused the cancellation
+
+				if !errors.Is(context.Cause(s.runningContext), restartNeededError) {
+					// this can only happen if the passed-in context is cancellable and is itself cancelled.
+					// In this case, the error, if any, is retrievable from the context and therefore should NOT be returned
+					// from this function.
+					debug("Start: context cancelled, cancellation from above, quitting")
+					return nil
+				}
+
+				// The conext has been cancelled and therefore the controller manager gracefully shut down.
+				// We actually waited for the start function to finish and cleared our running state.
+				// We also detected that the restart is needed by inspecting the cancellation cause.
+				//
+				// All the conditions for the restart are satisfied, so we can just start another loop that will start the manager
+				// anew.
+				debug("Start: context cancelled, restart request detected, entering another start loop")
+			}
+		case err := <-startErr:
+			debug("Start: error received, quitting")
+			return err
+		}
+	}
+}
+
+func (s *StartManager) doStart(ctx context.Context) (<-chan error, <-chan struct{}) {
+	debug("doStart: enter")
+	s.lock.Lock()
+	defer func() {
+		s.lock.Unlock()
+		debug("doStart: unlocked")
+	}()
+
+	debug("doStart: locked")
+
+	s.runningContext = nil
+	s.cancelFunc = nil
+
+	// we cannot guarantee the order in which the caller will wait for these channels vs when the channels are pushed to
+	// in this method (because we start the manager in a co-routine, or, actually, when the error is sent to before the
+	// control flow returns to the caller).
+	// We can guarantee, that the caller will eventually expect a message from these channels.
+	// Therefore, make the channels with a buffer of size 1, so that we can send a message here before the caller
+	// starts listening.
+	err := make(chan error, 1)
+	finished := make(chan struct{}, 1)
+
+	if s.running {
+		debug("doStart: already running")
+		err <- alreadyRunningError
+		return err, finished
+	}
+
+	s.runningContext, s.cancelFunc = context.WithCancelCause(ctx)
+
+	debug("doStart: calling InitializeManager")
+	mgr, initErr := s.InitializeManager()
+	if initErr != nil {
+		debug("doStart: InitializeManager failed")
+		err <- initErr
+		return err, finished
+	}
+	go func() {
+		debug("doStart: about to invoke mgr.Start()")
+		if startErr := mgr.Start(s.runningContext); startErr != nil {
+			debug("doStart: mgr.Start() returned error. Sending it.")
+			err <- startErr
```

Comment:
if the manager cannot be started, then we can just simply force-kill, right?
```suggestion
			os.Exit(1)
```
There is nothing we could try to recover. Or is there any other cleanup we would need to do in this case?

Reply:
I have one more note/question - I understand the usage of the channel in this case, but I don't understand why we need to use the channel for all other error cases returned from this method. Is it to only unify the error handling, or is there any other reason?

Reply:
I am not a fan of panicking in "library" code.

As for the error handling - yes, the error is always returned in the channel so that the caller doesn't have to have special logic for when the "setup" of the start fails as opposed to the failure that happens async-ly during the startup and lifetime of the controller manager.